### ISSUE TITLE
Reproducibility Q&A and Updates for EpiAgent Code

### ISSUE THREAD

**Issue by xy-chen16** [xy-chen16/EpiAgent#12]
We have opened this issue as a centralized place for discussions and questions related to reproducing the results reported in our EpiAgent paper.  We will continue to use this thread to share updates, clarifications, and responses to reproduction-related questions.

First, please note that we previously updated the cCRE_frequency. npy file.  However, the previous version was not correctly normalized by the total number of cells, making it incompatible with our global_TFIDF function.  This issue does not affect the model input itself, but it may affect downstream analyses that rely on TF-IDF values.  We have now corrected this issue and updated the cCRE_frequency. npy file accordingly.

Second, we have created a dedicated folder named reproducibility￼ to host code for reproducing the results in our paper.  At present, we have uploaded the code for reproducing the data imputation experiments.  Due to the scale issue in the previously updated cCRE_frequency. npy file, earlier attempts to reproduce the results may have been affected.  We apologize for the inconvenience.  In this update, we have corrected the scale of cCRE_frequency. npy and have rerun the full pipeline locally to ensure that the results can be reproduced.

Finally, for the inter-dataset annotation experiments, we also provide the raw data and processed labels used in our paper.  Users can directly access them through the following Google Drive link:

https://drive.google.com/drive/folders/16YZP-gdMkHm4vYSQLYUTGo1dTAwLc6n7?usp=share_link

Please feel free to leave questions or issues related to reproducibility in this thread.  We will do our best to respond and provide further clarification.

**Comment by galgantar** [comment]
Thank you for providing the reproducibility section, it is indeed very helpful. Additionally I am interested if you could expand the `data_imputation.py`script to also include the clustering experiment with ARI and NMI calculation on the Buenrostro2018 dataset (figure 2j from the original paper).

Thank you in advance.

**Comment by xy-chen16** [comment]
Thank you for your interest in the reproducibility materials.

I asked one of my collaborators to organize the relevant code for the clustering experiment. We have now added a script named `cluster_and_umap.py` to the `reproducibility` folder. It includes the clustering workflow and the calculation of ARI and NMI used in our experiments.

Hopefully this will help you reproduce the results in Figure 2j more easily. Please let us know if you have any further questions.

Best,
Xiaoyang

**Comment by galgantar** [comment]
Thank you for providing the script. I have a question regarding the clustering algorithm used. The `cluster_and_umap.py` uses Louvain clustering while the paper mentions that Leiden clustering was used.

Additionally, even after swapping `sc.tl.louvain` with `sc.tl.leiden` running `cluster_and_umap.py` on the outputs of the imputed signals from `data_imputation.py` for Buenrostro2018 dataset I get NMI=0.44 and ARI=0.24 when reported values from the paper are NMI=0.699 and ARI=0.544.

Thank you in advance.

**Comment by galgantar** [comment]
@xy-chen16  I would really appreciate your response on this when you have a chance, especially regarding the clustering method and the settings needed to reproduce the reported NMI and ARI values.

**Comment by xy-chen16** [comment]
Sorry for the long delay in getting back to you.

After seeing your previous comments, we went back and tried to reproduce the results ourselves, and we actually ran into the same issue. We’ve spent quite a while debugging it.

The data imputation experiments were originally conducted on a separate computing platform. Unfortunately, after the project on that platform expired, the corresponding files and scripts were removed. When we later organized the reproducibility code, we reused the training settings from some of our other experiments, which turned out to be incorrect for the data imputation task. This was why we were also unable to reproduce the original clustering results.

We tried many possible explanations (e.g., different TF-IDF scaling, different TF-IDF preprocessing methods, etc.), but none of them resolved the issue.

Recently, we managed to recover a backup of the original training code from our server, and to our surprise, we were able to reproduce the reported clustering performance. The key difference was actually very simple: we accidentally replaced the original learning rate scheduler when organizing the code.

Instead of

scheduler = LambdaLR(optimizer, lr_lambda=noam_scheduler)

the original implementation uses the following scheduler:

from torch.optim.lr_scheduler import _LRScheduler
class NoamLR(_LRScheduler):
    """
    Implements the Noam learning rate schedule.
    The learning rate increases linearly for the first `warmup_steps`
    training steps and then decreases proportionally to the inverse
    square root of the step number.
    """
    def __init__(self, optimizer, warmup_steps):
        self.warmup_steps = warmup_steps
        super().__init__(optimizer)
    def get_lr(self):
        step = max(1, self.last_epoch)
        scale = (
            self.warmup_steps ** 0.5
            * min(step ** (-0.5), step * self.warmup_steps ** (-1.5))
        )
        return [base_lr * scale for base_lr in self.base_lrs]
scheduler = NoamLR(optimizer, warmup_steps=warmup_steps)

After switching back to this scheduler, we were able to reproduce the clustering results.

This was our mistake when cleaning up and merging the training code for release—we incorrectly treated the two scheduler implementations as equivalent, but they are not.

We’ll update the reproducibility code in the repository soon. If you would like to reproduce the results before then, you can temporarily replace the scheduler with the implementation above.

Thank you again for your careful debugging and for bringing this issue to our attention.
